File: escala.py
import mydates as mydates

def gerarEscala(isoDate, listaPessoas):

    mesLista = mydates.gerarMesLista(isoDate)
    if not mesLista:
        return []
    newPessoasLista = listaPessoas.copy()
    newMes = []
    for semana in mesLista:
        newSemana = []
        for dia in range(7):
            if len(newPessoasLista) == 0:
                newPessoasLista = listaPessoas.copy()

            if semana[dia] != '0':
                # tenta gerar uma vez
                escalaDia = geraEscalaDia(dia, semana[dia], newPessoasLista)
                # caso nao funcione, extende a lista e tenta de novo
                if (escalaDia == {}):
                    newPessoasLista.extend(listaPessoas.copy())
                    escalaDia = geraEscalaDia(dia, semana[dia], newPessoasLista)
                # caso funcione faz o append, se nao ignora e vai para o proximo
                if (escalaDia != {}):
                    newSemana.append(escalaDia)

        newMes.append(newSemana)

    return newMes


def geraEscalaDia(diaIdx, diaString, listaPessoas):
    escalaDia = {}
    for pessoaIdx in range(len(listaPessoas)):
        pessoa = listaPessoas[pessoaIdx]
        diaHome = mydates.getDiaIndex(pessoa.homeOffice)
        if diaHome != diaIdx:
            escalaDia = { "dia": diaString,
                          "nome": pessoa.nome
                        }
            del listaPessoas[pessoaIdx]
            break
    return escalaDia

File: src/escala.js
import {gerarListaMes, defaultTimezone} from './myDates';

export default async function gerarEscala(dados){

    const listaMes = await gerarListaMes(dados);
    if ( listaMes === null ){
        return null;
    };

    const listaPessoas = dados.listaPessoas;
    const newListaMes = [];
    const pessoasDia = dados.pessoasDia;
    const diasOptions = dados.diasOptions;

    let newListaPessoas = [...listaPessoas];

    listaMes.forEach( (dia) => {

        const escalaDia = {
            dia: dia,
            nome: []
        };

        for (let i = 0; i < pessoasDia; i++){

            if (newListaPessoas.length == 0){
                newListaPessoas = [...listaPessoas];
            };

            let newEscalaDia = geraEscalaDia(dia, newListaPessoas, diasOptions);

            // caso nao funcione, extende a lista e tenta de novo
            if (newEscalaDia === null){
    
                newListaPessoas = [
                    ...newListaPessoas,
                    ...listaPessoas
                ];
                newEscalaDia = geraEscalaDia(dia, newListaPessoas, diasOptions);
            };

            // só acrescenta se achar alguem disponivel e ainda nao incluido no dia
            if (newEscalaDia !== null && !escalaDia.nome.includes(newEscalaDia.nome)){
                escalaDia.nome.push( newEscalaDia.nome );
            };
        };

        if (escalaDia.nome.length > 0) {
            newListaMes.push(escalaDia)
        };

    });

    return newListaMes;
};

function geraEscalaDia(dia, listaPessoas, diasOptions){

    let escalaDia = null;
    const d = new Date(dia+defaultTimezone);
    const diaIdx = d.getDay();

    for (let pessoaIdx = 0; pessoaIdx < listaPessoas.length; pessoaIdx++ ){
        const pessoa = listaPessoas[pessoaIdx];
        const pessoaListaSemanaIndex = pessoa.listaSemana.map( (diaSemana) => diasOptions[diaSemana] );
        // se o dia do mes e da semana não estiver nas listas, coloca a pessoa na escala e tira ela da fila
        if (!pessoaListaSemanaIndex.includes(diaIdx) && !pessoa.listaMes.includes(dia)){
            escalaDia = {   "dia": dia,
                            "nome": pessoa.nome
            };
            listaPessoas.splice(pessoaIdx, 1);
            break;
        };
    };

    return escalaDia;
};